lib/evse_security/evse_security.cpp

```diff
@@ -550,66 +533,23 @@ std::string EvseSecurity::generate_certificate_signing_request(LeafCertificateTy
-    EC_KEY* ecKey = EC_KEY_new_by_curve_name(NID_X9_62_prime256v1);
-    X509_NAME* x509Name = X509_REQ_get_subject_name(x509ReqPtr.get());
-
-    BIO_ptr prkey(BIO_new_file(key_path.c_str(), "w"));
-    BIO_ptr bio(BIO_new(BIO_s_mem()));
-
-    // generate ec key pair
-    EC_KEY_generate_key(ecKey);
-    EVP_PKEY_assign_EC_KEY(evpKey.get(), ecKey);
-    // write private key to file
-    int success;
-    if (this->private_key_password.has_value()) {
-        success = PEM_write_bio_PrivateKey(prkey.get(), evpKey.get(), EVP_aes_128_cbc(), NULL, 0, NULL,
-                                           (void*)this->private_key_password.value().c_str());
-    } else {
-        success = PEM_write_bio_PrivateKey(prkey.get(), evpKey.get(), NULL, NULL, 0, NULL, NULL);
-    }
-
-    if (!success) {
-        throw std::runtime_error("Failed to write private key");
-    }
-
-    // set version of x509 req
-    X509_REQ_set_version(x509ReqPtr.get(), n_version);
-
-    // set subject of x509 req
-    X509_NAME_add_entry_by_txt(x509Name, "C", MBSTRING_ASC, reinterpret_cast<const unsigned char*>(country.c_str()), -1,
-                               -1, 0);
-    X509_NAME_add_entry_by_txt(x509Name, "O", MBSTRING_ASC,
-                               reinterpret_cast<const unsigned char*>(organization.c_str()), -1, -1, 0);
-    X509_NAME_add_entry_by_txt(x509Name, "CN", MBSTRING_ASC, reinterpret_cast<const unsigned char*>(common.c_str()), -1,
-                               -1, 0);
-    X509_NAME_add_entry_by_txt(x509Name, "DC", MBSTRING_ASC, reinterpret_cast<const unsigned char*>("CPO"), -1, -1, 0);
-    // set public key of x509 req
-    X509_REQ_set_pubkey(x509ReqPtr.get(), evpKey.get());
+    std::string csr;
+    CertificateSigningRequestInfo info;
 
-    STACK_OF(X509_EXTENSION)* extensions = sk_X509_EXTENSION_new_null();
-    X509_EXTENSION* ext_key_usage = X509V3_EXT_conf_nid(NULL, NULL, NID_key_usage, "digitalSignature, keyAgreement");
-    X509_EXTENSION* ext_basic_constraints = X509V3_EXT_conf_nid(NULL, NULL, NID_basic_constraints, "critical,CA:false");
-    sk_X509_EXTENSION_push(extensions, ext_key_usage);
-    sk_X509_EXTENSION_push(extensions, ext_basic_constraints);
+    info.n_version = 0;
+    info.commonName = common;
+    info.country = country;
+    info.organization = organization;
 
-    X509_REQ_add_extensions(x509ReqPtr.get(), extensions);
+    info.private_key_file = key_path;
+    if (private_key_password.has_value())
+        info.private_key_pass = private_key_password;
 
-    // set sign key of x509 req
-    X509_REQ_sign(x509ReqPtr.get(), evpKey.get(), EVP_sha256());
-
-    // write csr
-    PEM_write_bio_X509_REQ(bio.get(), x509ReqPtr.get());
-
-    BUF_MEM* mem_csr = NULL;
-    BIO_get_mem_ptr(bio.get(), &mem_csr);
-    std::string csr(mem_csr->data, mem_csr->length);
+    if (false == CryptoSupplier::x509_generate_csr(info, csr)) {
+        throw std::runtime_error("Failed to generate certificate signing request!");
```

Comment:
Should this be a fatal error? Is it possible this is triggered by an incorrect request from the outside?

Reply:
From the old code, it was treated as a failure, for example when failing to write the private key (case in which the function fails). I am not entirely sure how this is supposed to be treaded, but nonetheless that's how it was treated.